### PR TITLE
Oserhiienko: #1730: filters panel

### PR DESCRIPTION
**Pull request by Aleksashka11**

### CODE REVIEW

File: packages/Bio/src/package.ts [hunk 1]
```diff
@@ -640,7 +640,7 @@ export function saveAsFasta() {
   saveAsFastaUI();
 }
 
-//name: BioSubstructureFilter
+//name: Bio Substructure Filter
```
Comment (skalkin) [inline packages/Bio/src/package.ts:643]: Why does this ScaffoldTree-focused PR have this Bio-related change?

File: packages/Chem/src/widgets/scaffold-tree.ts [hunk 2]
```diff
@@ -349,11 +348,21 @@ export class ScaffoldTreeViewer extends DG.JsViewer {
   Table: string;
   treeEncode: string;
 
+  sizesMap: {[key: string]: {[key: string]: number}} = {
```
Comment (skalkin) [inline packages/Chem/src/widgets/scaffold-tree.ts:351]: This is not typed well enough. Clearly you rely on instances having "height" and "width", this should be reflected.

File: packages/Chem/src/widgets/scaffold-tree.ts [hunk 3]
```diff
@@ -1136,9 +1154,21 @@ export class ScaffoldTreeViewer extends DG.JsViewer {
       this.updateFilters();
       this._bitOpInput!.value = this.bitOperation;
     }
+    else if (p.name === 'size') {
```
Comment (skalkin) [inline packages/Chem/src/widgets/scaffold-tree.ts:1157]: Why would be reconstruct the whole tree when the size is changed?

File: packages/Chem/src/widgets/scaffold-tree.ts [hunk 4]
```diff
@@ -1136,9 +1154,21 @@ export class ScaffoldTreeViewer extends DG.JsViewer {
...
+      ScaffoldTreeViewer.deserializeTrees(savedTree, this.tree, (molStr: string, rootGroup: TreeViewGroup) => {
+        return this.createGroup(molStr, rootGroup, false);
+      });
+      this.updateSizes();
+      this.updateUI();
+    }
   }
 
   onFrameAttached(dataFrame: DG.DataFrame): void {
+    while (this.root.firstChild)
```
Comment (skalkin) [inline packages/Chem/src/widgets/scaffold-tree.ts:1169]: There is a utility method for that, I think it's us.tools.empty, or something like that

File: packages/Chem/src/widgets/scaffold-tree.ts [hunk 5]
```diff
@@ -1156,6 +1186,9 @@ export class ScaffoldTreeViewer extends DG.JsViewer {
       if (orphans)
         return;
 
+      if (FLAG)
```
Comment (skalkin) [inline packages/Chem/src/widgets/scaffold-tree.ts:1189]: What is this "FLAG" (use more meaningful name) and why is it CAPITALIZED?

File: packages/Chem/src/widgets/scaffold-tree.ts [hunk 6]
```diff
@@ -1512,3 +1544,56 @@ class SketcherDialogWrapper {
...
+  savedTree: string = '';
+  
+  constructor() {
+    super();
+    this.root = ui.divV([]);
+    this.subs = [];
+  };
+
+  get isFiltering(): boolean {
+    return super.isFiltering;
```
Comment (skalkin) [inline packages/Chem/src/widgets/scaffold-tree.ts:1559]: isFiltering indicates whether the filter is filtering something (in this case, whether a user clicked on a substructure to filter on), or is it just in its default state. And anyway, why would you return super.isFiltering?

File: packages/Chem/src/widgets/scaffold-tree.ts [hunk 7]
```diff
@@ -1512,3 +1544,56 @@ class SketcherDialogWrapper {
...
+    super.attach(dataFrame);
+    this.column ??= dataFrame.columns.bySemType(DG.SEMTYPE.MOLECULE);
+    this.columnName ??= this.column?.name;
+    this.createViewer(dataFrame);
+  };
+
+  saveState(): any {
+    const state = super.saveState();
+    state.savedTree = JSON.stringify(ScaffoldTreeViewer.serializeTrees(this.viewer.tree));
+    this.viewer.updateFilters(this.isFiltering);
```
Comment (skalkin) [inline packages/Chem/src/widgets/scaffold-tree.ts:1576]: Why would you update filters when saving the state?

**Review by skalkin — CHANGES_REQUESTED** [review]

File: packages/Chem/src/widgets/scaffold-tree.ts (hunk 2 above)
Comment (Aleksashka11) [reply packages/Chem/src/widgets/scaffold-tree.ts]: Fixed

File: packages/Chem/src/widgets/scaffold-tree.ts (hunk 4 above)
Comment (Aleksashka11) [reply packages/Chem/src/widgets/scaffold-tree.ts]: Yes, I found ui.empty for such case. Fixed too

File: packages/Chem/src/widgets/scaffold-tree.ts (hunk 7 above)
Comment (Aleksashka11) [reply packages/Chem/src/widgets/scaffold-tree.ts]: Fixed

File: packages/Chem/src/widgets/scaffold-tree.ts (hunk 3 above)
Comment (Aleksashka11) [reply packages/Chem/src/widgets/scaffold-tree.ts:1173]: Fixed

File: packages/Chem/src/widgets/scaffold-tree.ts (hunk 6 above)
Comment (Aleksashka11) [reply packages/Chem/src/widgets/scaffold-tree.ts:1573]: isFiltering is needed here in order to check whether the filter has been disabled by the user. 
According to docs:  * This is used to minimize the number of unnecessary computations.
     * Make sure to call super.isFiltering to check whether the filter has been disabled by the user *
It is really useful, especially when scaffold tree viewer performs filtering itself, this is the way to control it. Same way our substructure filter works and others.

File: packages/Chem/src/widgets/scaffold-tree.ts [hunk 8]
```diff
@@ -1512,3 +1558,62 @@ class SketcherDialogWrapper {
...
+  attach(dataFrame: DG.DataFrame): void {
+    super.attach(dataFrame);
+    this.column ??= dataFrame.columns.bySemType(DG.SEMTYPE.MOLECULE);
+    this.columnName ??= this.column?.name;
+    this.createViewer(dataFrame);
+    this.subs.push(this.dataFrame!.onRowsFiltering
+      .pipe(filter((_) => !this.isFiltering))
+      .subscribe((_: any) => {
+        delete this.column!.temp['chem-scaffold-filter'];
+        setTimeout(() => this.viewer.updateFilters(this.isFiltering), 100);
```
[the remark this answers is not in the document]
Comment (Aleksashka11) [reply packages/Chem/src/widgets/scaffold-tree.ts:1589]: Sure, I'll fix that

[the remark this answers is not in the document]
Comment (Aleksashka11) [reply packages/Chem/src/widgets/scaffold-tree.ts]: Fixed

File: packages/Chem/src/widgets/scaffold-tree.ts [hunk 9]
```diff
@@ -1512,3 +1558,62 @@ class SketcherDialogWrapper {
...
+    return super.isFiltering;
+  }
+
+  get filterSummary(): string {
+    return ScaffoldTreeViewer.TYPE;  
+  }
+
+  attach(dataFrame: DG.DataFrame): void {
+    super.attach(dataFrame);
+    this.column ??= dataFrame.columns.bySemType(DG.SEMTYPE.MOLECULE);
```
[the remark this answers is not in the document]
Comment (skalkin) [reply packages/Chem/src/widgets/scaffold-tree.ts:1583]: Check RadioFilter or SubstructureFilter, they work. Or are you saying they don't? That would be a problem

[the remark this answers is not in the document]
Comment (skalkin) [reply packages/Chem/src/widgets/scaffold-tree.ts:1583]: What other filters have this problem? I'll fix it

**Review by skalkin — APPROVED** [review]

**Pull request merged**